app/services/slot_builder.rb

```diff
@@ -0,0 +1,175 @@
+# frozen_string_literal: true
+
+module SlotBuilder
+  # À faire avant, au moment de jouer avec le motifs
+  # @for_agents ? motifs : motifs.reservable_online
+  # Ce filtre est lié à la recherche de plage d'ouverture à partir d'un nom de motif... Est-ce vraiment nécessaire dans notre cas ?
+  #
+  # @for_agents sert aussi pour « limiter » l'afficahge des créneaux. Je pense que c'est à faire sur la vue.
+  # uniq_by = @for_agents ? ->(c) { [c.starts_at, c.agent_id] } : ->(c) { c.starts_at }
+  #  creneaux.uniq(&uniq_by).sort_by(&:starts_at)
+
+  class << self
+    # méthode publique
+    def available_slots(motif, lieu, date_range, off_days, agents = [])
+      datetime_range = ensure_date_range_with_time(date_range)
+      plage_ouvertures = plage_ouvertures_for(motif, lieu, datetime_range, agents)
+      free_times_po = free_times_from(plage_ouvertures, datetime_range, off_days) # dépendance sur RDV et Absence
+      slots_for(free_times_po, motif)
+    end
+
+    def ensure_date_range_with_time(date_range)
+      time_begin = date_range.begin.is_a?(Time) ? date_range.being : date_range.begin.beginning_of_day
+      time_begin = Time.zone.now if time_begin < Time.zone.now
+      time_end = date_range.end.is_a?(Time) ? date_range.end : date_range.end.beginning_of_day
+
+      time_begin..time_end
+    end
+
+    def plage_ouvertures_for(motif, lieu, datetime_range, agents)
+      lieu.plage_ouvertures.merge(motif.plage_ouvertures).not_expired.in_range(datetime_range)
+        .where(({ agent: agents } if agents&.any?))
```

Comment:
Suggestion de style: je trouverais ça plus lisible avec un critère par ligne, par exemple
```suggestion
      PlageOuverture.not_expired
        .merge(lieu.plage_ouvertures)
        .merge(motif.plage_ouvertures)
        .in_range(datetime_range)
        .where(({ agent: agents } if agents&.any?))
```

(D’ailleurs, je connaissais pas la formulation du dernier critère, qui revient à faire `where(nil)` si `agents` est vide.)